### ISSUE TITLE
Chaining subtasks - helpers for `chain` and `takeAndQueue`?

### ISSUE THREAD

**Issue by mxdubois** [yelouafi/redux-saga#174]
Hey @yelouafi. First of all, this library is really impressive. Thank you for creating and maintaining it. My team and I are excited about it and seriously considering using it.

I've been playing around a bit and it seems to me that the API is missing a convenient way to chain tasks sequentially in the background without blocking the current generator (sort of a saga analog to `promise.then`).

Imagine I want to ensure that all `FOO` actions that are dispatched are handled, but never in parallel. Essentially, I would want to do something like this (ideally with a little more sugar in the general case):

```javascript
function* fooWatcher() {
  let compositeTask = yield fork(noOp);
  while (true) {
    const action = yield take('FOO');
    compositeTask = yield fork(chain, compositeTask, call(handleFoo, action));
  }
}

// This is one possible implementation of chain.. .
// I go back and forth on whether or not the `fork` should be included in `chain` or not.
function* chain(task, effect) {
  yield join(task);
  yield effect;
}
```

I think the solution to this could entail both a generic helper for chaining background tasks in any situation and a `take`-based helper, to complement `takeEvery` -- something like:

```javascript
yield takeAndQueue('FOO', handleFoo);
```

So my questions to you...

1. Does the basic concept make sense?
2. Would you want to add something like this to the collection of helpers (whether they're in core or moved out of core)?
3. Am I on the right track for implementing `chain`, or would you go about it differently?

**Comment by joonhyublee** [comment]
+1 This would fit my use case too. Am using @mxdubois's implementation. How about 'takeSequentially' for the name? 'takeAndQueue' doesn't sound so bad either :)

**Comment by joonhyublee** [comment]
Since compositeTask persists forever with the saga itself, (unlike the promise chain) it might not get garbage collected and eat up memory as chain gets long, even after all the effects are resolved..?

**Comment by mxdubois** [comment]
My impression is that each call to `yield fork(...)` returns a new task
object. I haven't looked at whether or not that task object has any
references to the other task objects or any bearing over whether or not
those tasks are cleaned up as they complete.

**Comment by yelouafi** [comment]
@mxdubois 
>Does the basic concept make sense?

Yes it does. There are various concurrency patterns, and this one of them (takeEvery, takeLatest, takeFirst/takeAndQueue, takeWithConcurrencyLimit...). As an analogy you can look at the flatmap*** methods on Rx Obervables (flatmap, flatmapLatest, flatmapFirst ...)


>Would you want to add something like this to the collection of helpers (whether they're in core or moved out of core)?

I prefer to have this out of the core for now. Saga helpers are actually [implemented as state machines](https://github.com/yelouafi/redux-saga/blob/master/src/internal/sagaHelpers.js) (see #107 for the reason). Implementing advanced concurrency solutions using state machines will add more complexity to the code. I prefer to have implementations written using Generators instead.

Using Generators we can implement more sophisticated concurrency patterns. A suitable place for this is in the [Common Concurrency Patterns](http://yelouafi.github.io/redux-saga/docs/advanced/Concurrency.html) section of the docs. Later we can gather all the helper functions into a separate package (like `redux-saga-concurrency`) along with the tests.

**Comment by yelouafi** [comment]
>My impression is that each call to `yield fork(...)` returns a new task
object.

Yes, `fork` returns a [Task Object](http://yelouafi.github.io/redux-saga/docs/api/index.html#task)

>I haven't looked at whether or not that task object has any references to the other task objects or any bearing over whether or not those tasks are cleaned up as they complete.

a task will be normally cleaned up as soon as : 

1. it's completed (either success or error)
2. No external references (in the user code) are kept to it

**Comment by joonhyublee** [comment]
@mxdubois anyway, do you reckon `yield join(task)` is necessary here? Wouldn't `yield task` work too?
